nextcloud/server
# Fix status check and saving of external storages

This pull request fixes several issues with the status check in the external storage administration UI:
- Not showing it automatically for user-global storages
- Not removing it when the configuration of a storage is changed
- Not updating the tooltip in some cases when the status is updated
- Untranslated UI string

It also fixes some issues when saving an storage in the external storage administration UI:
- Not being able to save several storages of the same type in a row
- Saving `__unmodified__` as the password if it was not changed, which messes the saved configuration

Please refer to the individual commit descriptions for further details.

## How to test (scenario 1)

- Install sshd in the system running the Nextcloud server (in Debian/Ubuntu, `apt install openssh-server`)
- Add a user to the system (for example, `useradd --create-home sftptest`)
- Start the sshd daemon (if systemctl is available, `systemctl start sshd`; otherwise, for example, in Docker, `mkdir /run/sshd && /usr/sbin/sshd -D`)
- In Nextcloud, log in as an admin
- Open the Administration settings
- Open the external storage section
- Add a new SFTP backend
- Set authentication to `Manually entered, store in database`
- Set host to `127.0.0.1`
- Enable the storage for all users
- Save the storage
- Open the external storage section under the Personal settings
- Set the user and password for the storage
- Save the storage
- Reload the page

### Result with this pull request

The status check is automatically shown for the storage. Moreover, if the storage is saved again without changes the status will succeed again

### Result without this pull request

The status check is not automatically shown for the storage. Moreover, if the user saves the configuration again to trigger the status check (rather than just clicking on the status check) the status will fail (because the password is unmodified and obfuscated, and that obfuscated password is the one saved)



## How to test (scenario 2)

- Add a new external storage and fill it with the right configuration
- Save it; the status check will appear with a successful result
- Remove any of the fields, that should make the status check fail if triggered again

### Result with this pull request

It is not possible to click on the status check (until the configuration is saved and the status check is shown again)

### Result without this pull request

A click on the status check triggers a new check, which ends succeeding due to using the old configuration



## How to test (scenario 3)

- See steps in scenario 1 to setup SSH server
- Add a new SFTP backend
- Set host to `127.0.0.1`
- Set wrong credentials
- Enable the storage for all users
- Save the storage
- Hover on the error status icon

### Result with this pull request

Tooltip was updated with the error message (_Exception: Login failed_)

### Result without this pull request

Tooltip is still the original message (_Click to recheck the configuration_)



## How to test (scenario 4)

- See steps in scenario 1 to setup SSH server
- Add a new SFTP backend
- Set host to `127.0.0.1`
- Set wrong credentials
- Enable the storage for all users
- Save the storage
- Recheck the configuration by clicking on the status icon
- Fix the credentials
- Save the storage again
- Hover on the success status icon

### Result with this pull request

Tooltip was restored with the default message (_Click to recheck the configuration_)

### Result without this pull request

Tooltip is still the error message set when rechecking it (_Exception: Login failed_)



## How to test (scenario 5)

- Add a new external storage and save it
- Open the external storage settings in another tab and remove the storage in it
- In the original tab, try to remove the storage

### Result with this pull request

Tooltip was updated with the error message (_Storage with ID "X" not found_)

### Result without this pull request

Tooltip is still the original message (_Click to recheck the configuration_)



## How to test (scenario 6)

- Add a new external storage and save it
- Try to add another external storage of the same backend type

### Result with this pull request

_Add storage_ is selected in the dropdown, and selecting again the same backend type as before adds a new entry

### Result without this pull request

The same backend type is still selected in the dropdown, and selecting it again does nothing
